File: exporter/SynthesisFusionAddin/src/Parser/SynthesisParser/Components.py
# Contains all of the logic for mapping the Components / Occurrences
from platform import python_build

import adsk.core
import adsk.fusion
from google.protobuf.message import Error
from requests.models import parse_header_links

from src.ErrorHandling import Err, ErrorSeverity, Ok, Result
from src.Logging import getLogger, logFailure
from src.Parser.ExporterOptions import ExporterOptions
from src.Parser.SynthesisParser import PhysicalProperties
from src.Parser.SynthesisParser.PDMessage import PDMessage
from src.Parser.SynthesisParser.Utilities import (
    fill_info,
    guid_component,
    guid_occurrence,
)
from src.Proto import assembly_pb2, joint_pb2, material_pb2, types_pb2
from src.Types import ExportMode


# TODO: Impelement Material overrides
def mapAllComponents(
    design: adsk.fusion.Design,
    options: ExporterOptions,
    progressDialog: PDMessage,
    partsData: assembly_pb2.Parts,
    materials: material_pb2.Materials,
) -> Result[None]:

    for component in design.allComponents:
        adsk.doEvents()
        if progressDialog.wasCancelled():
            raise RuntimeError("User canceled export")
        progressDialog.addComponent(component.name)

        comp_ref = guid_component(component)

        fill_info_result = fill_info(partsData, None)
        if fill_info_result.is_err() and fill_info_result.unwrap_err()[1] == ErrorSeverity.Fatal:
            return fill_info_result

        partDefinition = partsData.part_definitions[comp_ref]

        fill_info_result = fill_info(partDefinition, component, comp_ref)
        if fill_info_result.is_err() and fill_info_result.unwrap_err()[1] == ErrorSeverity.Fatal:
            return fill_info_result

        physical_properties_result = PhysicalProperties.getPhysicalProperties(component, partDefinition.physical_data)
        if physical_properties_result.is_err() and physical_properties_result.unwrap_err()[1] == ErrorSeverity.Fatal:
            return physical_properties_result

        partDefinition.dynamic = options.exportMode != ExportMode.FIELD

        def processBody(body: adsk.fusion.BRepBody | adsk.fusion.MeshBody) -> Result[None]:
            if progressDialog.wasCancelled():
                raise RuntimeError("User canceled export")
            if body.isLightBulbOn:
                part_body = partDefinition.bodies.add()

                fill_info_result = fill_info(part_body, body)
                if fill_info_result.is_err() and fill_info_result.unwrap_err()[1] == ErrorSeverity.Fatal:
                    return fill_info_result

                part_body.part = comp_ref

                fill_info_result = fill_info(part_body, body)
                if fill_info_result.is_err():
                    return fill_info_result

                if isinstance(body, adsk.fusion.BRepBody):
                    parse_result = parseBRep(body, options, part_body.triangle_mesh)
                    if parse_result.is_err() and parse_result.unwrap_err()[1] == ErrorSeverity.Fatal:
                        return parse_result
                else:
                    parse_result = parseMesh(body, options, part_body.triangle_mesh)
                    if parse_result.is_err() and parse_result.unwrap_err()[1] == ErrorSeverity.Fatal:
                        return parse_result

                appearance_key = "{}_{}".format(body.appearance.name, body.appearance.id)
                # this should be appearance
                if appearance_key in materials.appearances:
                    part_body.appearance_override = appearance_key
                else:
                    part_body.appearance_override = "default"

            return Ok(None)

        for body in component.bRepBodies:
            process_result = processBody(body)
            if process_result.is_err() and process_result.unwrap_err()[1] == ErrorSeverity.Fatal:
                return process_result
        for body in component.meshBodies:
            process_result = processBody(body)
            if process_result.is_err() and process_result.unwrap_err()[1] == ErrorSeverity.Fatal:
                return process_result

    return Ok(None)


def parseComponentRoot(
    component: adsk.fusion.Component,
    progressDialog: PDMessage,
    options: ExporterOptions,
    partsData: assembly_pb2.Parts,
    material_map: dict[str, material_pb2.Appearance],
    node: types_pb2.Node,
) -> Result[None]:
    mapConstant = guid_component(component)

    part = partsData.part_instances[mapConstant]

    node.value = mapConstant

    fill_info_result = fill_info(part, component, mapConstant)
    if fill_info_result.is_err() and fill_info_result.unwrap_err()[1] == ErrorSeverity.Fatal:
        return fill_info_result

    def_map = partsData.part_definitions

    if mapConstant in def_map:
        part.part_definition_reference = mapConstant

    for occur in component.occurrences:
        if progressDialog.wasCancelled():
            raise RuntimeError("User canceled export")

        if occur.isLightBulbOn:
            child_node = types_pb2.Node()

            parse_child_result = parseChildOccurrence(
                occur, progressDialog, options, partsData, material_map, child_node
            )
            if parse_child_result.is_err():
                return parse_child_result

            node.children.append(child_node)
    return Ok(None)


def parseChildOccurrence(
    occurrence: adsk.fusion.Occurrence,
    progressDialog: PDMessage,
    options: ExporterOptions,
    partsData: assembly_pb2.Parts,
    material_map: dict[str, material_pb2.Appearance],
    node: types_pb2.Node,
) -> Result[None]:
    if occurrence.isLightBulbOn is False:
        return Ok(None)

    progressDialog.addOccurrence(occurrence.name)

    mapConstant = guid_occurrence(occurrence)

    compRef = guid_component(occurrence.component)

    part = partsData.part_instances[mapConstant]

    node.value = mapConstant

    fill_info_result = fill_info(part, occurrence, mapConstant)
    if fill_info_result.is_err() and fill_info_result.unwrap_err()[1] == ErrorSeverity.Fatal:
        return fill_info_result

    collision_attr = occurrence.attributes.itemByName("synthesis", "collision_off")
    if collision_attr != None:
        part.skip_collider = True

    if occurrence.appearance:
        try:
            part.appearance = "{}_{}".format(occurrence.appearance.name, occurrence.appearance.id)
        except:
            _: Err[None] = Err("Failed to format part appearance", ErrorSeverity.Warning)
            part.appearance = "default"
        # TODO: Add phyical_material parser

    # TODO: I'm fairly sure that this should be a fatal error
    if occurrence.component.material:
        part.physical_material = occurrence.component.material.id
    else:
        return Err(f"Component Material is None", ErrorSeverity.Fatal)

    def_map = partsData.part_definitions

    if compRef in def_map:
        part.part_definition_reference = compRef

    # TODO: Maybe make this a separate step where you dont go backwards and search for the gamepieces
    if options.exportMode == ExportMode.FIELD:
        for x in options.gamepieces:
            if x.occurrenceToken == mapConstant:
                partsData.part_definitions[part.part_definition_reference].dynamic = True
                break

    part.transform.spatial_matrix.extend(occurrence.transform.asArray())

    worldTransform = getMatrixWorld(occurrence)

    if worldTransform:
        part.global_transform.spatial_matrix.extend(worldTransform.asArray())

    for occur in occurrence.childOccurrences:
        if progressDialog.wasCancelled():
            raise RuntimeError("User canceled export")

        if occur.isLightBulbOn:
            child_node = types_pb2.Node()

            parse_child_result = parseChildOccurrence(
                occur, progressDialog, options, partsData, material_map, child_node
            )
            if parse_child_result.is_err():
                return parse_child_result

            node.children.append(child_node)
    return Ok(None)


# saw online someone used this to get the correct context but oh boy does it look pricey
# I think if I can make all parts relative to a parent it should return that parents transform maybe
# TESTED AND VERIFIED - but unoptimized
def getMatrixWorld(occurrence: adsk.fusion.Occurrence) -> adsk.core.Matrix3D:
    matrix = occurrence.transform2
    while occurrence.assemblyContext:
        matrix.transformBy(occurrence.assemblyContext.transform2)
        occurrence = occurrence.assemblyContext
    return matrix


def parseBRep(
    body: adsk.fusion.BRepBody,
    options: ExporterOptions,
    trimesh: assembly_pb2.TriangleMesh,
) -> Result[None]:

    calc = body.meshManager.createMeshCalculator()
    # Disabling for now. We need the user to be able to adjust this, otherwise it gets locked
    # into whatever the default was at the time it first creates the export options.
    # calc.setQuality(options.visualQuality)
    _ = calc.setQuality(adsk.fusion.TriangleMeshQualityOptions.LowQualityTriangleMesh)
    # calc.maxNormalDeviation = 3.14159 * (1.0 / 6.0)
    # calc.surfaceTolerance = 0.5
    try:
        mesh = calc.calculate()
    except:
        return Err(f"Failed to calculate mesh for {body.name}", ErrorSeverity.Error)

    fill_info_result = fill_info(trimesh, body)
    if fill_info_result.is_err() and fill_info_result.unwrap_err()[1] == ErrorSeverity.Fatal:
        return fill_info_result

    trimesh.has_volume = True

    plainmesh_out = trimesh.mesh
    plainmesh_out.verts.extend(mesh.nodeCoordinatesAsFloat)
    plainmesh_out.normals.extend(mesh.normalVectorsAsFloat)
    plainmesh_out.indices.extend(mesh.nodeIndices)
    plainmesh_out.uv.extend(mesh.textureCoordinatesAsFloat)

    return Ok(None)


def parseMesh(
    meshBody: adsk.fusion.MeshBody,
    options: ExporterOptions,
    trimesh: assembly_pb2.TriangleMesh,
) -> Result[None]:
    mesh = meshBody.displayMesh
    if mesh is None:
        return Err("Component Mesh was None", ErrorSeverity.Fatal)

    fill_info_result = fill_info(trimesh, meshBody)
    if fill_info_result.is_err() and fill_info_result.unwrap_err()[1] == ErrorSeverity.Fatal:
        return fill_info_result

    trimesh.has_volume = True

    plainmesh_out = trimesh.mesh

    plainmesh_out.verts.extend(mesh.nodeCoordinatesAsFloat)
    plainmesh_out.normals.extend(mesh.normalVectorsAsFloat)
    plainmesh_out.indices.extend(mesh.nodeIndices)
    plainmesh_out.uv.extend(mesh.textureCoordinatesAsFloat)

    return Ok(None)


def mapRigidGroups(rootComponent: adsk.fusion.Component, joints: joint_pb2.Joints) -> None:
    groups = rootComponent.allRigidGroups
    for group in groups:
        mira_group = joint_pb2.RigidGroup()
        mira_group.name = group.entityToken
        for occ in group.occurrences:
            if occ == None:
                a = 1
                continue

            if not occ.isLightBulbOn:
                continue

            occRef = guid_occurrence(occ)
            mira_group.occurrences.append(occRef)
        if len(mira_group.occurrences) > 1:
            joints.rigid_groups.append(mira_group)

File: isotope/src/components.cpp
#include "components.h"

#include <Core/Geometry/Matrix3D.h>

#include <vector>

#include "assembly.pb.h"
#include "joint.pb.h"
#include "types.pb.h"

#include "util.h"

namespace {

mirabuf::PhysicalProperties map_physical_properties(
    const adsk::core::Ptr<adsk::fusion::PhysicalProperties>& properties) {
    mirabuf::PhysicalProperties new_properties;
    new_properties.set_mass(properties->mass());
    new_properties.set_volume(properties->volume());
    new_properties.set_density(properties->density());
    new_properties.set_area(properties->area());
    if (auto com = properties->centerOfMass()) {
        if (auto vec = com->asVector()) {
            new_properties.mutable_com()->set_x(vec->x());
            new_properties.mutable_com()->set_y(vec->y());
            new_properties.mutable_com()->set_z(vec->z());
        }
    }

    return new_properties;
}

mirabuf::TriangleMesh map_b_rep_body(const adsk::core::Ptr<adsk::fusion::BRepBody>& body) {
    auto mesh_mgr = body->meshManager();
    if (!mesh_mgr) {
        return {};
    }

    auto calc = mesh_mgr->createMeshCalculator();
    if (!calc) {
        return {};
    }

    calc->setQuality(adsk::fusion::TriangleMeshQualityOptions::LowQualityTriangleMesh);
    auto fus_mesh = calc->calculate();
    if (!fus_mesh) {
        return {};
    }

    mirabuf::TriangleMesh mesh;
    mesh.mutable_info()->CopyFrom(create_info_from_fus_obj(body));

    mesh.set_has_volume(true);

    std::vector<float> coords = fus_mesh->nodeCoordinatesAsFloat();
    mesh.mutable_mesh()->mutable_verts()->Add(coords.begin(), coords.end());

    std::vector<float> normals = fus_mesh->normalVectorsAsFloat();
    mesh.mutable_mesh()->mutable_normals()->Add(normals.begin(), normals.end());

    std::vector<int> node_indicies = fus_mesh->nodeIndices();
    mesh.mutable_mesh()->mutable_indices()->Add(node_indicies.begin(), node_indicies.end());

    std::vector<float> texture_coords = fus_mesh->textureCoordinatesAsFloat();
    mesh.mutable_mesh()->mutable_uv()->Add(texture_coords.begin(), texture_coords.end());

    return mesh;
}

mirabuf::TriangleMesh map_mesh_body(const adsk::core::Ptr<adsk::fusion::MeshBody>& body) {
    auto fus_mesh = body->displayMesh();
    if (!fus_mesh) {
        return {};
    }

    mirabuf::TriangleMesh mesh;
    mesh.mutable_info()->CopyFrom(create_info_from_fus_obj(body));
    mesh.set_has_volume(true);

    std::vector<float> coords = fus_mesh->nodeCoordinatesAsFloat();
    mesh.mutable_mesh()->mutable_verts()->Add(coords.begin(), coords.end());

    std::vector<float> normals = fus_mesh->normalVectorsAsFloat();
    mesh.mutable_mesh()->mutable_normals()->Add(normals.begin(), normals.end());

    std::vector<int> node_indicies = fus_mesh->nodeIndices();
    mesh.mutable_mesh()->mutable_indices()->Add(node_indicies.begin(), node_indicies.end());

    std::vector<float> texture_coords = fus_mesh->textureCoordinatesAsFloat();
    mesh.mutable_mesh()->mutable_uv()->Add(texture_coords.begin(), texture_coords.end());

    return mesh;
}

adsk::core::Ptr<adsk::core::Matrix3D> get_matrix_world(const adsk::core::Ptr<adsk::fusion::Occurrence>& occurrence) {
    if (!occurrence) {
        return nullptr;
    }

    auto transform = occurrence->transform2();
    if (!transform) {
        return nullptr;
    }

    auto matrix          = transform->copy();
    auto next_occurrence = occurrence;
    while (next_occurrence->assemblyContext()) {
        next_occurrence     = next_occurrence->assemblyContext();
        auto next_transform = next_occurrence->transform2();
        if (!next_transform) {
            break;
        }

        matrix->transformBy(next_transform);
    }

    return matrix;
}

mirabuf::Node parse_child_occurrence(
    const adsk::core::Ptr<adsk::fusion::Occurrence>& occurrence, mirabuf::Parts* parts) {
    if (!occurrence->isLightBulbOn()) {
        return {};
    }

    mirabuf::Node node;

    // TODO: Really explicit typing for this sort of thing would be great.
    const std::string map_constant = guid_occurrence(occurrence);
    node.set_value(map_constant);

    if (parts->part_instances().find(map_constant) != parts->part_instances().end()) {
        return node;
    }

    auto& part = (*parts->mutable_part_instances())[map_constant];
    part.mutable_info()->CopyFrom(create_info_from_fus_obj(occurrence, map_constant));
    if (occurrence->appearance()) {
        part.set_appearance(occurrence->appearance()->id());
    } else {
        part.set_appearance("default");
    }

    auto component = occurrence->component();
    if (component) {
        if (auto material = component->material()) {
            part.set_physical_material(material->id());
        }

        auto& part_defs                 = parts->part_definitions();
        const std::string component_ref = guid_component(component);
        if (part_defs.find(component_ref) != part_defs.end()) {
            part.set_part_definition_reference(component_ref);
        }
    }

    auto transform_array = occurrence->transform()->asArray();
    part.mutable_transform()->mutable_spatial_matrix()->Add(transform_array.begin(), transform_array.end());

    if (auto world_matrix = get_matrix_world(occurrence)) {
        auto world_transform = world_matrix->asArray();
        part.mutable_global_transform()->mutable_spatial_matrix()->Add(world_transform.begin(), world_transform.end());
    }

    // final recursive step to parse child occurrences
    std::vector<adsk::core::Ptr<adsk::fusion::Occurrence>> child_occurrences;
    if (auto occurrence_list = occurrence->childOccurrences()) {
        occurrence_list->copyTo(std::back_inserter(child_occurrences));
    }
    for (const auto& child_occurrence : child_occurrences) {
        if (!child_occurrence->isLightBulbOn()) {
            continue;
        }

        auto child_node = parse_child_occurrence(child_occurrence, parts);
        node.mutable_children()->Add()->CopyFrom(child_node);
    }

    return node;
}

mirabuf::Parts build_part_definitions(const adsk::core::Ptr<adsk::fusion::Components>& components,
    const google::protobuf::Map<std::string, mirabuf::material::Appearance>& appearances) {
    mirabuf::Parts parts;

    if (!components) {
        return parts;
    }

    std::vector<adsk::core::Ptr<adsk::fusion::Component>> fusion_components;
    components->copyTo(std::back_inserter(fusion_components));
    for (const auto& component : fusion_components) {
        const std::string component_ref = guid_component(component);
        if (parts.part_definitions().find(component_ref) != parts.part_definitions().end()) {
            continue;
        }

        auto& part = (*parts.mutable_part_definitions())[component_ref];
        part.mutable_info()->CopyFrom(create_info_from_fus_obj(component, component_ref));
        part.set_dynamic(true);

        if (auto props = component->physicalProperties()) {
            *part.mutable_physical_data() = map_physical_properties(props);
        }

        std::vector<adsk::core::Ptr<adsk::fusion::BRepBody>> b_rep_bodies;
        if (auto body_list = component->bRepBodies()) {
            body_list->copyTo(std::back_inserter(b_rep_bodies));
        }
        for (const auto& body : b_rep_bodies) {
            if (!body->isLightBulbOn()) {
                continue;
            }

            auto& part_body = *part.mutable_bodies()->Add();
            part_body.mutable_info()->CopyFrom(create_info_from_fus_obj(body));
            part_body.mutable_triangle_mesh()->CopyFrom(map_b_rep_body(body));

            auto appearance = body->appearance();
            if (appearance && appearances.find(appearance->id()) != appearances.end()) {
                part_body.set_appearance_override(appearance->id());
            } else {
                part_body.set_appearance_override("default");
            }
        }

        std::vector<adsk::core::Ptr<adsk::fusion::MeshBody>> mesh_bodies;
        if (auto mesh_body_list = component->meshBodies()) {
            mesh_body_list->copyTo(std::back_inserter(mesh_bodies));
        }
        for (const auto& body : mesh_bodies) {
            if (!body->isLightBulbOn()) {
                continue;
            }

            auto& part_body = *part.mutable_bodies()->Add();
            part_body.mutable_info()->CopyFrom(create_info_from_fus_obj(body));
            part_body.mutable_triangle_mesh()->CopyFrom(map_mesh_body(body));

            auto appearance = body->appearance();
            if (appearance && appearances.find(appearance->id()) != appearances.end()) {
                part_body.set_appearance_override(appearance->id());
            } else {
                part_body.set_appearance_override("default");
            }
        }
    }

    return parts;
}

mirabuf::Node parse_component_root(const adsk::core::Ptr<adsk::fusion::Component>& component, mirabuf::Parts* parts) {
    mirabuf::Node root_node;
    const std::string map_constant = guid_component(component);
    root_node.set_value(map_constant);

    if (parts->part_instances().find(map_constant) != parts->part_instances().end()) {
        return root_node;
    }

    auto& part = (*parts->mutable_part_instances())[map_constant];
    part.mutable_info()->CopyFrom(create_info_from_fus_obj(component, map_constant));
    auto& part_defs = parts->part_definitions();
    if (part_defs.find(map_constant) != part_defs.end()) {
        part.set_part_definition_reference(map_constant);
    }

    std::vector<adsk::core::Ptr<adsk::fusion::Occurrence>> child_occurrences;
    if (auto occurrence_list = component->occurrences()) {
        occurrence_list->copyTo(std::back_inserter(child_occurrences));
    }
    for (const auto& child_occurrence : child_occurrences) {
        if (!child_occurrence->isLightBulbOn()) {
            continue;
        }

        auto child_node = parse_child_occurrence(child_occurrence, parts);
        root_node.mutable_children()->Add()->CopyFrom(child_node);
    }

    return root_node;
}

} // namespace

std::pair<mirabuf::Parts, mirabuf::Node> map_parts(const adsk::core::Ptr<adsk::fusion::Components>& components,
    const adsk::core::Ptr<adsk::fusion::Component>& root, const mirabuf::material::Materials& materials) {
    auto parts     = build_part_definitions(components, materials.appearances());
    auto root_node = parse_component_root(root, &parts);
    return {std::move(parts), std::move(root_node)};
}

void map_rigid_groups(const adsk::core::Ptr<adsk::fusion::Component>& root, mirabuf::joint::Joints* joints) {
    for (const auto& fus_group : root->allRigidGroups()) {
        if (!fus_group) {
            continue;
        }

        auto mira_group = mirabuf::joint::RigidGroup();
        mira_group.set_name(fus_group->entityToken());

        auto occurrences = fus_group->occurrences();
        if (!occurrences) {
            continue;
        }

        for (const auto& occurrence : occurrences) {
            if (!occurrence || !occurrence->isLightBulbOn()) {
                continue;
            }

            mira_group.mutable_occurrences()->Add(guid_occurrence(occurrence));
        }

        if (mira_group.occurrences().size() > 1) {
            joints->mutable_rigid_groups()->Add()->CopyFrom(mira_group);
        }
    }
}
